### ISSUE TITLE
misleading "Extra inputs are not permitted" error on dataclass `Field` with `pattern` where type is a union with a `list`

### ISSUE THREAD

**Issue by DetachHead** [pydantic/pydantic#6577]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

```
pydantic_core._pydantic_core.SchemaError: Invalid Schema:
dataclass.schema.dataclass-args.fields.0.schema.default.schema.nullable.schema.list.pattern
  Extra inputs are not permitted [type=extra_forbidden, input_value='.*', input_type=str]
    For further information visit https://errors.pydantic.dev/2.1.2/v/extra_forbidden
```

this error is misleading, i thought it was caused by `{"extra": "forbid"}`

### Example Code

```Python
from pydantic import Field
from pydantic.dataclasses import dataclass


@dataclass
class Foo:
    a: list[str] | None = Field(pattern=".*", default=None)
```

### Python, Pydantic & OS Version

```Text
pydantic version: 2.0.2
        pydantic-core version: 2.1.2 release build profile
                 install path: C:\Users\user\AppData\Local\pypoetry\Cache\virtualenvs\project-zbC6VfKB-py3.11\Lib\site-packages\pydantic
               python version: 3.11.2 (tags/v3.11.2:878ead1, Feb  7 2023, 16:38:35) [MSC v.1934 64 bit (AMD64)]
                     platform: Windows-10-10.0.19044-SP0
     optional deps. installed: ['devtools', 'typing-extensions']
```

Selected Assignee: @hramezani

**Comment by hramezani** [comment]
Thanks @DetachHead for reporting this 🙏 

I can confirm this.

Actually, you cannot apply `patter` to `list[str]`. It means if you change the field definition to
`a: list[str] = Field(pattern=".*")`
you will get 
`TypeError: The following constraints cannot be applied to list[str]: 'pattern'`

**Comment by DetachHead** [comment]
yeah i stumbled across this incorrect code when updating to pydantic 2, i guess pydantic 1 ignored the problem.

is there an actual way to apply regex patterns to items in a `list`?

**Comment by hramezani** [comment]
You can use `AfterValidator` like:

```py
import re
from typing import List

from typing_extensions import Annotated

from pydantic import AfterValidator, BaseModel


def check_pattern(data: str) -> str:
    assert re.match('test\d', data) is not None
    return data

MyStr = Annotated[str, AfterValidator(check_pattern)]


class Model(BaseModel):
    a: List[MyStr]


Model(a=['test1', 'test2', ''])

```

**Comment by samuelcolvin** [comment]
@DetachHead I agree the error is a little confusing, but technically it's correct.

Here's why:
* in this case you set `pattern` in `Field`, but the type is `list[str] | None` - a union, so what pydantic is naively doing is adding `"patern": ".*"` to the `Union` core-schema definition
* pydantic-core uses pydantic-core to perform validation on pydantic-core schema before constructing the validator and serializer
* It's this validation of the core-schema which is seeing an extra item in the union type-dict  and raising an error

In V1 we sometimes did something we called "magical pushdown" - if the field type didn't accept a constraint (e.g. `pattern`) we looked for a nested type within that type that did. This worked sometimes but could be confusing. Hence we removed it in V2.

I'm not really sure how to go about making this error more readable, some options:
* do some checks in pydantic before adding these constraints - and raise an error if we see an incorrect constraints - this wouldn't need to be perfect as the core-schema validation would catch any rarer cases
* add some more explanation to the errors raised by core-schema validation explaining what's going on
* use a different validation error in core-schema for `extra-forbidden` - something like "the core schema got an unexpected item 'pattern', you might be applying a constraint to a type where it doesn't make sense"

LMK what you think?

**Comment by DetachHead** [comment]
any of those options sound good to me

**Comment by sydney-runkle** [comment]
We've attempted to improve this by moving the error to runtime. Now you get:

```py
TypeError: Unable to apply constraint 'pattern' to supplied value ['a', 'b'] for schema of type 'list'
```

Marking as resolved, accordingly! Releasing v2.9 shortly, there'll be some commentary in the associated blog post re this change.
